### PR TITLE
Variant implementation & integration into x509::Certificate

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: include/aos/common/tools/variant.hpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
```
Comment (al1img) [inline include/aos/common/tools/variant.hpp:2]: 2024

File: include/aos/common/tools/variant.hpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_VARIANT_HPP_
+#define AOS_VARIANT_HPP_
+
+#include "buffer.hpp"
```
Comment (al1img) [inline include/aos/common/tools/variant.hpp:11]: Looks like not used.

File: include/aos/common/tools/variant.hpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+template <typename T, typename U, typename... Ts>
+struct GetTypeIndex<T, U, Ts...> {
+    static constexpr int Value = 1 + GetTypeIndex<T, Ts...>::Value;
+};
+
+/**
+ * Base static visitor for the variant.
+ */
+
```
Comment (al1img) [inline include/aos/common/tools/variant.hpp:39]: Remove empty line.

File: include/aos/common/tools/variant.hpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    {
+        if (typeInd == 0) {
+            auto& val = StaticCast(variant.GetPtr());
+            return visitor.Visit(val);
+        } else {
+            return VisitorHelper<VarArgs...>::ApplyVisitor(typeInd - 1, visitor, variant);
+        }
+    }
+
+    static const T& StaticCast(const void* ptr) { return *static_cast<const T*>(ptr); }
```
Comment (al1img) [inline include/aos/common/tools/variant.hpp:77]: Private? Is it needed at all? Could we use RemoveConst?

**Review by MykolaSolyanko — APPROVED** [review]

**Review by al1img — APPROVED** [review]

**Pull request merged**